pkg/storegateway/series_chunks.go

```diff
@@ -417,6 +422,183 @@ func (c *loadingSeriesChunksSetIterator) Err() error {
+		numRanges := len(s.chunksRanges)
+		partialSeries[i].refsRanges = s.chunksRanges
+		partialSeries[i].rawRanges = make([][]byte, numRanges)
+		chunksCount := 0
+		for _, g := range s.chunksRanges {
+			chunksCount += len(g.refs)
+		}
+		partialSeries[i].parsedChunks = getPooledChunks(chunksCount)
+	}
+	return partialSeries
+}
+
+//nolint:unused // dead code while we are working on PR 3968
+type underfetchedChunksRangeIdx struct {
+	blockID  ulid.ULID
+	rangeIdx int
+	parsed   []storepb.AggrChunk
+}
+
+// parse tries to parse the ranges in the partial set with the bytes it has in rawRanges.
+// If the bytes for any of the ranges aren't enough, then parse will return an underfetchedChunksRangeIdx
+// and will correct the length of all ranges which had a understimated length.
+// Currently, parse will only correct the length of the last chunk, since this is the only chunk
+// whose size we estimate.
+// parse will also truncate the bytes in rawRanges in case there are extra unnecessary bytes there.
+// parse will return an error when the bytes in rawRanges were underfetched before the last chunk or
+// if the data in rawRanges is invalid.
+//
+//nolint:unused // dead code while we are working on PR 3968
+func (s partialSeriesChunksSet) parse() ([]underfetchedChunksRangeIdx, error) {
+	var underfetchedRanges []underfetchedChunksRangeIdx
+	parsedChunksCount := 0
+	for rIdx, r := range s.refsRanges {
+		dst := s.parsedChunks[parsedChunksCount : parsedChunksCount+len(r.refs)]
+		ok, err := s.populateRange(rIdx, dst)
+		if err != nil {
+			return nil, fmt.Errorf("parsing chunk range (block %s, first ref %d, num chunks %d): %w", r.blockID, r.firstRef(), len(r.refs), err)
+		}
+		if !ok {
+			underfetchedRanges = append(underfetchedRanges, underfetchedChunksRangeIdx{
+				blockID:  r.blockID,
+				rangeIdx: rIdx,
+				parsed:   dst,
+			})
+		}
+
+		parsedChunksCount += len(r.refs)
+	}
+	return underfetchedRanges, nil
+}
+
+//nolint:unused // dead code while we are working on PR 3968
+func convertChunkEncoding(storageEncoding chunkenc.Encoding) (storepb.Chunk_Encoding, bool) {
+	switch storageEncoding {
+	case chunkenc.EncXOR:
+		return storepb.Chunk_XOR, true
+	case chunkenc.EncHistogram:
+		return storepb.Chunk_Histogram, true
+	case chunkenc.EncFloatHistogram:
+		return storepb.Chunk_FloatHistogram, true
+	default:
+		return 0, false
+	}
+}
+
+//nolint:unused // dead code while we are working on PR 3968
+func (s partialSeriesChunksSet) reparse(idx underfetchedChunksRangeIdx) error {
+	refsRange := s.refsRanges[idx.rangeIdx]
+	ok, err := s.populateRange(idx.rangeIdx, idx.parsed)
+	if err != nil {
+		return fmt.Errorf("parsing underfetched range (block %s first ref %d): %w", idx.blockID, refsRange.firstRef(), err)
+	}
+	if !ok {
+		return fmt.Errorf("chunk length doesn't match after refetching (first ref %d)", refsRange.firstRef())
+	}
+	return nil
+}
+
+// parseRange also corrects the length of the last chunk to the size that it actually is.
```

Comment:
> parseRange

[nit] The function is called `populateRange`. I would also improve a bit this function comment. It's not clear what it does just reading the function comment.